File: bindings/rust/tachyon/src/star.rs
use std::ptr::NonNull;

use tachyon_sys::{
    tachyon_msg_view_t, tachyon_star_acquire_tx, tachyon_star_commit, tachyon_star_commit_tx,
    tachyon_star_destroy, tachyon_star_flush, tachyon_star_get_state, tachyon_star_n_spokes,
    tachyon_star_poll, tachyon_star_rollback_tx, tachyon_star_t,
    tachyon_state_t_TACHYON_STATE_UNKNOWN,
};

use crate::Bus;
use crate::error::{TachyonError, from_raw};

pub struct StarBus {
    ptr: NonNull<tachyon_star_t>,
}

unsafe impl Send for StarBus {}

impl Drop for StarBus {
    fn drop(&mut self) {
        unsafe {
            tachyon_star_destroy(self.ptr.as_ptr());
        }
    }
}

impl StarBus {
    /// Create a star from `buses`.
    ///
    /// `node_ids`, if provided, must have the same length as `buses`.
    /// A negative value at index `i` means "no NUMA binding for spoke i".
    ///
    /// Internally calls `tachyon_bus_ref` on every bus before returning,
    /// so the caller's `Bus` handles may be dropped independently.
    pub fn create(buses: &[&Bus], node_ids: Option<&[i32]>) -> Result<Self, TachyonError> {
        if let Some(ids) = node_ids {
            if ids.len() != buses.len() {
                return Err(TachyonError::InvalidSize);
            }
        }

        let mut raw: Vec<*mut tachyon_sys::tachyon_bus_t> =
            buses.iter().map(|b| b.as_ptr()).collect();
        let node_ids_ptr = node_ids.map_or(std::ptr::null(), |ids| ids.as_ptr());
        let mut out = std::ptr::null_mut();
        let err = unsafe {
            tachyon_sys::tachyon_star_create(raw.as_mut_ptr(), raw.len(), node_ids_ptr, &mut out)
        };

        from_raw(err)?;
        Ok(Self {
            ptr: NonNull::new(out).expect("tachyon_star_create succeeded but returned null"),
        })
    }

    /// Drain up to `max_total` messages across all spokes in a round-robin.
    ///
    /// Returns a [`StarPollGuard`] that borrows `self`.  The guard commits
    /// (advances all spoke tails) when dropped, so callers must finish
    /// reading message data before dropping it.
    pub fn poll(&self, max_total: usize, budget_us: u64) -> StarPollGuard<'_> {
        let mut views: Vec<tachyon_msg_view_t> = (0..max_total)
            .map(|_| unsafe { std::mem::zeroed() })
            .collect();
        let mut spoke_indices: Vec<usize> = vec![0; max_total];

        let n = unsafe {
            tachyon_star_poll(
                self.ptr.as_ptr(),
                views.as_mut_ptr(),
                max_total,
                budget_us,
                spoke_indices.as_mut_ptr(),
            )
        };

        views.truncate(n);
        spoke_indices.truncate(n);

        StarPollGuard {
            star: self,
            views,
            spoke_indices,
            committed: false,
        }
    }

    /// Acquire a TX slot on `spoke_idx`.
    ///
    /// Returns `None` if the ring is full or `spoke_idx` is out of bounds.
    /// The returned [`StarTxGuard`] rolls back the slot on drop if not committed.
    pub fn acquire_tx(&self, spoke_idx: usize, max_size: usize) -> Option<StarTxGuard<'_>> {
        let raw = unsafe { tachyon_star_acquire_tx(self.ptr.as_ptr(), spoke_idx, max_size) };
        NonNull::new(raw.cast::<u8>()).map(|ptr| StarTxGuard {
            star: self,
            spoke_idx,
            ptr,
            max_size,
            committed: false,
        })
    }

    /// Flush pending TX data on `spoke_idx` to the consumer.
    pub fn flush(&self, spoke_idx: usize) {
        unsafe { tachyon_star_flush(self.ptr.as_ptr(), spoke_idx) }
    }

    /// Number of spokes.
    pub fn n_spokes(&self) -> usize {
        unsafe { tachyon_star_n_spokes(self.ptr.as_ptr()) }
    }

    /// State of spoke `spoke_idx`.  Returns `UNKNOWN` if out of bounds.
    pub fn get_state(&self, spoke_idx: usize) -> tachyon_sys::tachyon_state_t {
        if spoke_idx >= self.n_spokes() {
            return tachyon_state_t_TACHYON_STATE_UNKNOWN;
        }
        unsafe { tachyon_star_get_state(self.ptr.as_ptr(), spoke_idx) }
    }
}

/// RAII handle returned by [`StarBus::poll`].
///
/// Holds the views and spoke indices for the current poll batch.
/// Commits (advances all spoke ring-buffer tails) when dropped.
/// Call [`commit`](StarPollGuard::commit) explicitly to handle errors.
pub struct StarPollGuard<'a> {
    star: &'a StarBus,
    views: Vec<tachyon_msg_view_t>,
    spoke_indices: Vec<usize>,
    committed: bool,
}

impl StarPollGuard<'_> {
    /// Number of messages received.
    pub fn len(&self) -> usize {
        self.views.len()
    }

    /// True if no messages were received.
    pub fn is_empty(&self) -> bool {
        self.views.is_empty()
    }

    /// Iterate over received messages, each paired with its spoke index.
    pub fn iter(&self) -> impl Iterator<Item = StarMsgView<'_>> {
        self.views
            .iter()
            .zip(self.spoke_indices.iter())
            .map(|(v, &s)| StarMsgView {
                view: v,
                spoke_idx: s,
            })
    }

    /// Advance all spoke ring-buffer tails (release slots to producers).
    ///
    /// Must be called after the caller has finished reading all message data,
    /// the pointers inside each [`StarMsgView`] point into the ring buffer.
    pub fn commit(mut self) -> Result<(), TachyonError> {
        self.committed = true;
        from_raw(unsafe { tachyon_star_commit(self.star.ptr.as_ptr()) })
    }
}

impl Drop for StarPollGuard<'_> {
    fn drop(&mut self) {
        if !self.committed {
            unsafe { tachyon_star_commit(self.star.ptr.as_ptr()) };
        }
    }
}

/// A single message from a poll batch, with its originating spoke index.
pub struct StarMsgView<'a> {
    view: &'a tachyon_msg_view_t,
    spoke_idx: usize,
}

impl StarMsgView<'_> {
    /// Zero-copy slice into the ring buffer payload.
    ///
    /// Valid only until the parent [`StarPollGuard`] is committed or dropped.
    pub fn data(&self) -> &[u8] {
        if self.view.ptr.is_null() || self.view.actual_size == 0 {
            return &[];
        }
        unsafe { std::slice::from_raw_parts(self.view.ptr.cast::<u8>(), self.view.actual_size) }
    }

    /// Raw type_id (use `route_id()` / `msg_type()` to decode).
    pub fn type_id(&self) -> u32 {
        self.view.type_id
    }

    /// Spoke (bus) index this message was received on.
    pub fn spoke_idx(&self) -> usize {
        self.spoke_idx
    }

    /// Actual payload size in bytes.
    pub fn actual_size(&self) -> usize {
        self.view.actual_size
    }
}

/// RAII TX slot returned by [`StarBus::acquire_tx`].
///
/// Rolls back the slot (releases it without publishing) on drop if
/// [`commit`](StarTxGuard::commit) was not called.
pub struct StarTxGuard<'a> {
    star: &'a StarBus,
    spoke_idx: usize,
    ptr: NonNull<u8>,
    max_size: usize,
    committed: bool,
}

impl StarTxGuard<'_> {
    /// Copy `data` into the slot.  Panics if `data.len() > max_size`.
    pub fn write(&self, data: &[u8]) {
        assert!(
            data.len() <= self.max_size,
            "write: data ({}) exceeds max_size ({})",
            data.len(),
            self.max_size
        );
        unsafe { std::ptr::copy_nonoverlapping(data.as_ptr(), self.ptr.as_ptr(), data.len()) }
    }

    /// Mutable slice over the full reserved slot (length = `max_size`).
    pub fn as_mut_slice(&mut self) -> &mut [u8] {
        unsafe { std::slice::from_raw_parts_mut(self.ptr.as_ptr(), self.max_size) }
    }

    /// Publish `actual_size` bytes with `type_id` to the spoke.
    /// Flushes immediately (consumer-visible after this call).
    pub fn commit(mut self, actual_size: usize, type_id: u32) -> Result<(), TachyonError> {
        assert!(actual_size <= self.max_size);
        self.committed = true;
        from_raw(unsafe {
            tachyon_star_commit_tx(self.star.ptr.as_ptr(), self.spoke_idx, actual_size, type_id)
        })
    }
}

impl Drop for StarTxGuard<'_> {
    fn drop(&mut self) {
        if !self.committed {
            unsafe { tachyon_star_rollback_tx(self.star.ptr.as_ptr(), self.spoke_idx) };
        }
    }
}

File: bindings/js/src/ts/star.ts
import { createRequire } from 'node:module';
import { isMainThread } from 'node:worker_threads';

const _require = createRequire(import.meta.url);

interface NativeStarPollItem {
	data: Buffer;
	typeId: number;
	actualSize: number;
	spokeIdx: number;
}

interface NativeStarBinding {
	close(): void;
	poll(maxTotal: number, budgetUs: number): NativeStarPollItem[];
	commit(): void;
	acquireTx(spokeIdx: number, maxSize: number): Buffer | null;
	commitTx(spokeIdx: number, actualSize: number, typeId: number): void;
	rollbackTx(spokeIdx: number): void;
	flush(spokeIdx: number): void;
	getState(spokeIdx: number): number;
	nSpokes(): number;
}

interface NativeStarModule {
	TachyonStarBusNode: {
		create(buses: object[], nodeIds?: number[] | null): NativeStarBinding;
	};
}

function loadNative(): NativeStarModule {
	const candidates = [
		new URL('../build/Release/tachyon_node.node', import.meta.url).pathname,
		new URL('../build/Debug/tachyon_node.node', import.meta.url).pathname,
		new URL('../../build/Release/tachyon_node.node', import.meta.url).pathname,
		new URL('../../build/Debug/tachyon_node.node', import.meta.url).pathname,
	];

	for (const p of candidates) {
		try {
			return _require(p) as NativeStarModule;
		} catch {}
	}

	throw new Error(
		'tachyon_node.node not found. Run `npm run build:native` first.\n' + `Searched: ${candidates.join(', ')}`,
	);
}

const native = loadNative();

function warnMainThread(method: string): void {
	if (isMainThread) {
		console.warn(
			`[tachyon] StarBus.${method}() called on the main thread. ` +
				'Blocking poll calls will saturate the event loop. ' +
				'Consider moving IPC to a Worker.',
		);
	}
}

/**
 * One message received from a specific spoke during a {@link StarPollGuard} poll.
 */
export interface StarMsgView {
	readonly data: Buffer;
	readonly typeId: number;
	readonly actualSize: number;
	readonly spokeIdx: number;
}

/**
 * Holds the zero-copy batch returned by {@link StarBus.poll}.
 *
 * All {@link StarMsgView.data} buffers point directly into shared memory and are
 * invalidated on {@link commit}. Commit releases the ring-buffer slots across
 * all polled spokes.
 *
 * Use `using` or call {@link commit} explicitly before the next {@link StarBus.poll}.
 *
 * @example
 * ```ts
 * using guard = star.poll(64, 5_000n);
 * for (const msg of guard.messages) {
 *   process(msg.data, msg.spokeIdx);
 * }
 * // guard auto-commits on scope exit
 * ```
 */
export class StarPollGuard implements Disposable {
	readonly #handle: NativeStarBinding;
	readonly #messages: StarMsgView[];
	#committed = false;

	/** @internal */
	public constructor(handle: NativeStarBinding, messages: StarMsgView[]) {
		this.#handle = handle;
		this.#messages = messages;
	}

	/** All messages in this poll batch. */
	public get messages(): readonly StarMsgView[] {
		return this.#messages;
	}

	/** Number of messages in this batch. */
	public count(): number {
		return this.#messages.length;
	}

	/** `true` if no messages were drained. */
	public isEmpty(): boolean {
		return this.#messages.length === 0;
	}

	/**
	 * Advances consumer tails for all polled spokes and invalidates all
	 * {@link StarMsgView.data} buffers (ArrayBuffers are detached).
	 */
	public commit(): void {
		if (this.#committed) return;
		this.#committed = true;
		if (this.#messages.length > 0) {
			this.#handle.commit();
		}
	}

	/** Called automatically by the `using` keyword. */
	public [Symbol.dispose](): void {
		this.commit();
	}
}

/**
 * Holds an exclusive zero-copy TX slot in the producer arena of a specific spoke.
 * Write the payload directly into {@link slot}, then call {@link commit} or
 * {@link rollback}. `using` rolls back automatically if not yet committed.
 * {@link commit} calls `tachyon_star_commit_tx`, which flushes internally.
 *
 * @example
 * ```ts
 * using tx = star.acquireTx(0, 8);
 * if (tx) {
 *   tx.slot.writeBigUInt64BE(value, 0);
 *   tx.commit(8, MSG_TYPE);
 * }
 * ```
 */
export class StarTxGuard implements Disposable {
	readonly #handle: NativeStarBinding;
	readonly #spokeIdx: number;
	readonly #slot: Buffer;
	#consumed = false;

	/** @internal */
	public constructor(handle: NativeStarBinding, spokeIdx: number, slot: Buffer) {
		this.#handle = handle;
		this.#spokeIdx = spokeIdx;
		this.#slot = slot;
	}

	/**
	 * Writable buffer pointing directly into shared memory.
	 * Invalid after {@link commit} or {@link rollback}.
	 *
	 * @throws {Error} If the guard has already been consumed.
	 */
	public get slot(): Buffer {
		if (this.#consumed) throw new Error('StarTxGuard: slot has already been committed or rolled back.');
		return this.#slot;
	}

	/**
	 * Publishes `actualSize` bytes with `typeId` and flushes the spoke arena.
	 *
	 * @param actualSize Number of bytes written. Must not exceed the reservation.
	 * @param typeId     User-defined protocol identifier.
	 * @throws {Error} If the guard has already been consumed.
	 */
	public commit(actualSize: number, typeId: number): void {
		if (this.#consumed) throw new Error('StarTxGuard: slot has already been committed or rolled back.');
		this.#consumed = true;
		this.#handle.commitTx(this.#spokeIdx, actualSize, typeId);
	}

	/** Aborts the TX slot without publishing. No-op if already consumed. */
	public rollback(): void {
		if (this.#consumed) return;
		this.#consumed = true;
		this.#handle.rollbackTx(this.#spokeIdx);
	}

	/** Calls {@link rollback} if not yet consumed. */
	public [Symbol.dispose](): void {
		this.rollback();
	}
}

/**
 * Aggregates N independent SPSC arenas into a single round-robin polling loop
 * bounded by a TSC-calibrated time budget. One consumer per `StarBus`; one
 * producer per spoke.
 *
 * Each spoke is a {@link Bus} created by the producer (Listener) side. The star
 * holds Connector-side handles and ref-counts them internally; the caller may
 * close its own handles after {@link create} returns.
 *
 * All polling and TX operations must be driven from a single consumer thread.
 * {@link close} / `using` are safe from any thread.
 *
 * @example
 * ```ts
 * // Worker thread
 * using star = StarBus.create([c0, c1, c2]);
 * for (;;) {
 *   using guard = star.poll(64, 5_000n);
 *   if (!guard.isEmpty()) {
 *     for (const msg of guard.messages) process(msg);
 *   }
 * }
 * ```
 */
export class StarBus implements Disposable {
	readonly #handle: NativeStarBinding;
	#closed = false;

	private constructor(handle: NativeStarBinding) {
		this.#handle = handle;
	}

	/**
	 * Creates a {@link StarBus} from the given connector-side buses.
	 *
	 * Each bus is ref-counted internally; the caller may close its own handles
	 * after this returns.
	 *
	 * @param buses   Non-empty array of connector-side {@link Bus} instances.
	 * @param nodeIds Optional NUMA node IDs, same length as `buses`.
	 *                Negative values skip binding for that spoke.
	 *                `null` / omitted disables NUMA binding entirely.
	 * @throws {TypeError}  An element of `buses` is not a valid open bus handle.
	 * @throws {Error}      `buses` is empty or `nodeIds` length mismatches.
	 */
	public static create(buses: { _handle: object }[], nodeIds?: number[] | null): StarBus {
		if (buses.length === 0) {
			throw new Error('StarBus.create: buses must not be empty.');
		}

		if (nodeIds !== null && nodeIds !== undefined && nodeIds.length !== buses.length) {
			throw new Error('StarBus.create: nodeIds.length must equal buses.length.');
		}

		const rawHandles = buses.map((b, _) => {
			return b._handle;
		});

		const handle = native.TachyonStarBusNode.create(rawHandles, nodeIds ?? null);
		return new StarBus(handle);
	}

	/** Number of spokes. */
	public get nSpokes(): number {
		this.#assertOpen();
		return this.#handle.nSpokes();
	}

	/**
	 * Raw `tachyon_state_t` integer for `spokeIdx`.
	 * Returns `TACHYON_STATE_UNKNOWN` (5) if `spokeIdx` is out of range.
	 */
	public getState(spokeIdx: number): number {
		this.#assertOpen();
		return this.#handle.getState(spokeIdx);
	}

	/**
	 * Drains up to `maxTotal` messages across all spokes within `budgetUs`
	 * microseconds. Returns an empty guard if the budget expires before any
	 * message arrives.
	 *
	 * All {@link StarMsgView.data} buffers are zero-copy and SHM-backed.
	 *
	 * The returned guard must be committed or closed before the next `poll`.
	 *
	 * @param maxTotal  Upper bound on messages to drain. Must be > 0.
	 * @param budgetUs  TSC-bounded polling budget in microseconds.
	 */
	public poll(maxTotal: number, budgetUs: number): StarPollGuard {
		this.#assertOpen();
		warnMainThread('poll');
		const raw = this.#handle.poll(maxTotal, budgetUs);
		return new StarPollGuard(this.#handle, raw);
	}

	/**
	 * Acquires an exclusive zero-copy TX slot on `spokeIdx`.
	 * Returns `null` if the ring is full or `spokeIdx` is out of range.
	 *
	 * Prefer the guard pattern; call {@link StarTxGuard.commit} or let `using`
	 * roll back automatically.
	 *
	 * @param spokeIdx Zero-based spoke index.
	 * @param maxSize  Required contiguous byte capacity.
	 */
	public acquireTx(spokeIdx: number, maxSize: number): StarTxGuard | null {
		this.#assertOpen();
		const slot = this.#handle.acquireTx(spokeIdx, maxSize);
		if (slot === null) return null;
		return new StarTxGuard(this.#handle, spokeIdx, slot);
	}

	/**
	 * Notifies sleeping consumers on `spokeIdx` via a futex wake-up signal.
	 * Not needed after {@link StarTxGuard.commit}, which flushes internally.
	 */
	public flush(spokeIdx: number): void {
		this.#assertOpen();
		this.#handle.flush(spokeIdx);
	}

	/** Closes the bus and releases all internal references. Safe to call multiple times. */
	public close(): void {
		if (this.#closed) return;
		this.#closed = true;
		this.#handle.close();
	}

	/** Called automatically by the `using` keyword. */
	public [Symbol.dispose](): void {
		this.close();
	}

	#assertOpen(): void {
		if (this.#closed) throw new Error('StarBus: this bus has been closed.');
	}
}
